pandas/io/common.py

```diff
@@ -877,7 +877,8 @@ def read(self, size: int = -1) -> str | bytes:
         if self.decode:
             # memory mapping is applied before compression. Encoding should
             # be applied to the de-compressed data.
-            return content.decode(self.encoding, errors=self.errors)
+            final: bool = len(content) == 0
+            return self.decoder.decode(content, final=final)
```

Comment:
I like your simple fix.

I think you will not need the `bool` annotation. 

If `read` is never called like `read(size=-1)` (or a large chunk size to read the entire file) with no further call to `read` after it, then your solution works perfectly (if that's the case, please add a comment about this assumption). If that is not the case, then `read(size=-1)` would silently drop any multi-byte character that is cut-off at the end.

Reply:
How about
```py
final = size == -1 or len(content) < size
```
that should cover more cases. It doesn't cover the case, when the file has exactly `size` bytes left and no further `read` call happens afterwards (I don't think this happens).